Fix and refactor acts as favorable and acts as watchable

toy opened the pull request:
Fixed:

- `Favorable::Registry` was copied without changing the module inside from `Watchable` and method name from `"acts_as_watchable"`
- Registries were using `Watchable` to check that only models that have `acts_as_watchable` called in them were allowed, but `Watchable` is included in `ApplicationRecord`, so `Watchable::InstanceMethods` should be used instead. Also the `acts_as_watchable` was registering model before including the module
- Both registries were storing model classes, so in development new versions of those models were added when reloaded, but the registry still returned the first model class that was registered

Refactored:

- ~~No need to explicitly call `Registry.add(Model)` in initialisers, it is enough to just reference models to trigger autoload~~
- Renamed `Routes` to `RouteConstraint` for clarity
- Extract all methods from registries to `RegistryMethods`

toy committed:
fix copy pasting in OpenProject::Acts::Favorable::Registry

toy committed:
check for inclusion of right module for acts_as_favorable/acts_as_watchable

Scope module is included in ApplicationRecord, so all models havel it
included, but we want to check that only models that have acts_as_...
method called on them are added to the registry, so we should use
InstanceMethods instead.

toy committed:
find instance method module and acts_as_ method name dynamically

To prevent copy pasting prbolem

toy committed:
remove unneeded namespaces and add to registry at the right point

toy committed:
just mention acts_as_favorable/acts_as_watchable models instead of explicitly adding to registry

toy committed:
use to_prepare instead of after_initialize in acts_as_favorable/acts_as_watchable initializers

to_prepare is also called every time code is reloaded

toy committed:
no need to explicitly require parts of watchable, as lib_static is in auto load once

toy committed:
remove unused exists? methods of registries

toy committed:
remove unused models accessor from Acts::Watchable::Routes

toy committed:
extract RegistryMethods

toy committed:
make models registry method private

toy committed:
rename Routes to RouteConstraint

toy committed:
refactor spec for Acts::Watchable::RouteConstraint and copy for Acts::Favorable::RouteConstraint

Can't come up with a good name for shared module to extract code from
RouteConstraint modules and specs

toy committed:
cleaner heredoc in acts_as_watchable shared spec

toy committed:
simplify check for project association for acts_as_watchable

toy committed:
store class names instead of classes in acts_as_ registries

Storing classes in development may lead to problems when reloading
classes, as new versions of the class will be appended to the list, but
`instance` method will still find the old versions of the classes.

I considered adding caching for production environment, but additional
code is not justified by 3 times faster execution.

```ruby
require "benchmark/ips"

h = {"Project" => Project}

Benchmark.ips do |x|
  x.report("constantize") { "Project".constantize }

  x.report("hash access") { h["Project"] }

  x.compare!
end
```

```
Warming up --------------------------------------
         constantize     1.163M i/100ms
         hash access     3.397M i/100ms
Calculating -------------------------------------
         constantize     11.842M (± 1.5%) i/s -     59.291M in
5.007752s
         hash access     34.379M (± 1.3%) i/s -    173.239M in
5.039989s

Comparison:
         hash access: 34378546.0 i/s
         constantize: 11842325.0 i/s - 2.90x  slower
```

toy committed:
spec for RegistryMethods

toy commented on config/initializers/acts_as_watchable.rb:
```diff
@@ -1,8 +1,13 @@
-    .add(WorkPackage, Message, Forum, News, Meeting, Wiki, WikiPage)
+Rails.application.config.to_prepare do
+  Forum
+  Meeting
```
This model is coming from meeting module, so it seems more appropriate to move this line to `modules/meeting/lib/open_project/meeting/engine.rb`, but all modules are loaded through `Gemfile.modules`, so not sure what is better - keeping it here or moving it there.

klaustopher commented on config/initializers/acts_as_watchable.rb:
```diff
@@ -1,8 +1,13 @@
+  Meeting
+  Message
+  News
+  Wiki
+  WikiPage
+  WorkPackage
```
I find it very confusing that just mentioning a class name is having a side effect of a registry being filled. The previous version had the advantage that we explicitly write in the initializer what it does instead of just putting a class name in there. I'm not a big fan of this change, but if others can agree with this, I don't have a complaint

oliverguenther replied:
I guess this is due to this: 

> Both registries were storing model classes, so in development new versions of those models were added when reloaded, but the registry still returned the first model class that was registered

But this could be fixed in the to_prepare call to simply clear out the registry

oliverguenther replied:
I too prefer adding it explicitly

klaustopher commented on lib_static/open_project/acts/registry_methods.rb:
```diff
@@ -0,0 +1,58 @@
+            raise ArgumentError.new("Model #{model} does not include #{acts_as_method_name}")
+          end
+
+          class_names << model.name
```
Why did we change the behavior here to put the class names (as string) in the set and not the classes anymore? This requires us to use `constantize` in the `instance` method. When we still put the class names in the set, we can use the same `models.detect` method we used in the old implementation.

oliverguenther replied:
I guess this is due to this: 

> Both registries were storing model classes, so in development new versions of those models were added when reloaded, but the registry still returned the first model class that was registered

But this could be fixed in the to_prepare call to simply clear out the registry.

toy replied:
So you prefer a separate method to clear registry (or an argument of `add` method) over storing names of models? I'll have a look myself if it looks better

toy replied:
Using Hash both resolves model reloading problem and doesn't need walking the set using `detect`, but also added explicit resetting of registries in `to_prepare`

oliverguenther commented:
> No need to explicitly call Registry.add(Model) in initialisers, it is enough to just reference models to trigger autoload

On the surface, it is arguably the exact same thing albeit a bit more explicit. I prefer having some notion of what classes have it rather than relying on autoloading, as developers might simply forget it. With the registry, this could not be the case (as soon as they started using/testing the feature, the controllers were checking the registry to see which classes existed)

toy committed:
bring back using registries explicitly for clarity

toy committed:
use hash mapping name to class instead of storing just names

This will alternatively resolve the problem with duplicate classes in
development on reload without the need to explicitly reset the registry

toy committed:
explicitly reset registries on reload
